### PR TITLE
OAK-10765 : fetch document with only min required fields to speed up …

### PR DESCRIPTION
**Pull request by rishabhdaim**
…verfication of orphan nodes

### CODE REVIEW

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupport.java [hunk 1]
```diff
@@ -226,6 +236,46 @@ public Optional<NodeDocument> getOldestModifiedDoc(final Clock clock) {
...
+                doc.deepCopy(newDoc);
+                newDoc.keySet().retainAll(projectedSet);
+                return of(newDoc);
+            }
+
+        } finally {
+            Utils.closeIfCloseable(docs);
+        }
+
+        LOG.info("No Doc has been found with id [{}]", id);
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGCSupport.java]: Shouldn't this be eg `debug`?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoVersionGCSupport.java [hunk 2]
```diff
@@ -165,6 +166,27 @@ public Iterable<NodeDocument> getModifiedDocs(final long fromModified, final lon
...
+        if (fields != null && !fields.isEmpty()) {
+            result.projection(include(fields));
+        }
+
+        try(MongoCursor<BasicDBObject> cur = result.iterator()) {
+            return cur.hasNext() ? ofNullable(store.convertFromDBObject(NODES, cur.next())) : empty();
+        } catch (Exception ex) {
+            LOG.error("getDocument() <- error while fetching data from Mongo", ex);
+        }
+        LOG.info("No Doc has been found with id [{}], retuning empty", id);
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoVersionGCSupport.java]: same as the other, shouldn't this be eg `debug`?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 3]
```diff
@@ -1216,34 +1220,27 @@ private boolean isDeletedOrOrphanedNode(final NodeState traversedState,
...
             }
-            if (detailedGcMode == DetailedGCMode.GAP_ORPHANS
-                    || detailedGcMode == DetailedGCMode.GAP_ORPHANS_EMPTYPROPS) {
+            if (detailedGcMode == DetailedGCMode.GAP_ORPHANS || detailedGcMode == DetailedGCMode.GAP_ORPHANS_EMPTYPROPS) {
                 // check the ancestor docs for gaps
                 final Path docPath = doc.getPath();
-                final Path geaPath = greatestExistingAncestorOrSelf;
-                final Path geaChildPath = docPath
-                        .getAncestor(docPath.getDepth() - geaPath.getDepth() - 1);
+                final Path geaChildPath = docPath.getAncestor(docPath.getDepth() - greatestExistingAncestorOrSelf.getDepth() - 1);
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: no strong opinion, but why was this changed?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 4]
```diff
@@ -1216,34 +1220,27 @@ private boolean isDeletedOrOrphanedNode(final NodeState traversedState,
...
                 Boolean missingType = missingDocsTypes.get(geaChildPath);
                 if (missingType == null) {
                     // we don't have it cached yet - so do the potentially expensive find
-                    final NodeDocument d = nodeStore.getDocumentStore().find(NODES,
-                            Utils.getIdFromPath(geaChildPath));
-                    final boolean parentDocExists = d != null;
-                    missingType = !parentDocExists;
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1233]: no strong opinion, but why was this changed?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java [hunk 5]
```diff
@@ -1216,34 +1220,27 @@ private boolean isDeletedOrOrphanedNode(final NodeState traversedState,
...
-                            log.error("isDeletedOrOrphanedNode : knownNullDocs removal failed, size was {}",
-                                    missingDocsTypes.size());
+                            log.error("isDeletedOrOrphanedNode : knownNullDocs removal failed, size was {}", missingDocsTypes.size());
                             missingDocsTypes.clear();
                         }
                     }
                     missingDocsTypes.put(geaChildPath, missingType);
                 }
-                if (missingType == false) {
+                if (!missingType) {
```
Comment (stefan-egli) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: reason for the non typical notation was that `missingType` had 3 states: null/true/false, and it being false was one of the types.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 3 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: IntelliJ was suggesting to inline the variable.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 4 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java:1233]: removed the inline variables as suggested by intelliJ

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java (hunk 5 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java]: missingType can't be null here, `null` is handled in the above code. Again suggestion by intelliJ

**Review by stefan-egli — APPROVED** [review]
